tools/SourceKit/lib/SwiftLang/SwiftSourceDocInfo.cpp

```diff
@@ -1214,57 +1272,93 @@ static bool addCursorInfoForLiteral(
 
-  CursorSymbolInfo &MainSymbol = Data.Symbols.emplace_back();
   // The primary result for constructor calls, eg. `MyType()` should be
   // the type itself, rather than the constructor. The constructor will be
   // added as a secondary result.
-  if (auto Err =
-          fillSymbolInfo(MainSymbol, MainInfo, Info->getLoc(), AddSymbolGraph,
-                         Lang, Invoc, PreviousSnaps, Data.Allocator)) {
-    llvm::handleAllErrors(std::move(Err), [&](const llvm::StringError &E) {
-      Diagnostic = E.message();
-    });
+  if (!addSymbolInfo(MainInfo, false)) {
     return false;
   }
 
   if (MainInfo.VD != OrigInfo.VD && !OrigInfo.Unavailable) {
-    CursorSymbolInfo &CtorSymbol = Data.Symbols.emplace_back();
-    if (auto Err =
-            fillSymbolInfo(CtorSymbol, OrigInfo, Info->getLoc(), AddSymbolGraph,
-                           Lang, Invoc, PreviousSnaps, Data.Allocator)) {
-      // Ignore but make sure to remove the partially-filled symbol
-      llvm::handleAllErrors(std::move(Err), [](const llvm::StringError &E) {});
-      Data.Symbols.pop_back();
-    }
+    addSymbolInfo(OrigInfo, true);
   }
 
   // Add in shadowed declarations if on a decl. For references just go to the
   // actual declaration.
   if (!Info->isRef()) {
     for (auto D : Info->getShorthandShadowedDecls()) {
-      CursorSymbolInfo &SymbolInfo = Data.Symbols.emplace_back();
-      DeclInfo DInfo(D, Type(), /*IsRef=*/true, /*IsDynamic=*/false,
-                     ArrayRef<NominalTypeDecl *>(), Invoc);
-      if (auto Err =
-              fillSymbolInfo(SymbolInfo, DInfo, Info->getLoc(), AddSymbolGraph,
-                             Lang, Invoc, PreviousSnaps, Data.Allocator)) {
-        // Ignore but make sure to remove the partially-filled symbol
-        llvm::handleAllErrors(std::move(Err),
-                              [](const llvm::StringError &E) {});
-        Data.Symbols.pop_back();
+      DeclInfo DInfo(D, SubstitutionMap(), Type(), /*IsRef=*/true,
+                     /*IsDynamic=*/false, ArrayRef<NominalTypeDecl *>(), Invoc);
+      addSymbolInfo(DInfo, true);
+    }
+  }
+
+  // Add Symbol infos for the substitutions. Add each type only once.
+  if (ExpandSubstitutions) {
+    auto Params =
+        Info->getSubstitutionMap().getGenericSignature().getGenericParams();
+    llvm::DenseSet<CanType> Visited;
+    for (auto *GenericTy : Params) {
+      // Find the replacement type for this generic type
+      QuerySubstitutionMap GetType{Info->getSubstitutionMap()};
+      Type Ty = GetType(GenericTy);
+
+      // Add Symbol info for the replacement type
+      if (!Visited.contains(Ty->getCanonicalType())) {
+        DeclInfo DInfo(ide::getDeclFromType(Ty), SubstitutionMap(), Type(),
+                       true, false, {}, Invoc);
+        addSymbolInfo(DInfo, true);
+        Visited.insert(Ty->getCanonicalType());
+      }
+
+      // Add Symbol info for the generic type
+      if (!Visited.contains(GenericTy->getCanonicalType())) {
```

Comment:
These will always be unique. However I'm not sure what value there is in collecting information about the generic parameters referenced here. This doesn't depend on the substitutions; the generic parameters are from the generic signature of the generic declaration we're referencing